### PR TITLE
Standardize function signature of compare_images.

### PR DESCRIPTION
**Pull request by mkcor**
## Description

This PR is a follow-up to #7255.

## Checklist

- A descriptive but concise pull request title
- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [Unit tests](https://scikit-image.org/docs/dev/development/contribute.html#testing)
- A gallery example in `./doc/examples` for new features
- [Contribution guide](https://scikit-image.org/docs/dev/development/contribute.html) is followed

## Release note

Summarize the introduced changes in the code block below in one or a few sentences. The
summary will be included in the next release notes automatically:

```release-note
In `skimage.util.compare_images`, deprecate the parameter `image2`.
Instead use `image0`, `image1` to pass the compared images. Furthermore,
all other parameters will be turned into keyword-only parameters once
the deprecation is complete.
```

### CODE REVIEW

File: skimage/util/compare.py
Comment (lagru) [inline skimage/util/compare.py:13]: While using `image0=None` has the advantage of being simple, I'd prefer a wrapper around `compare_images` here. That way, we can still perform the necessary checks, but `functools.wraps(func)` should ensure that IDE's and code completion pick up the actual signature I'm suggesting below
```suggestion
    image0, image1, image2=DEPRECATED, method='diff', *, n_tiles=(8, 8)
```
Note you can't yet move the `*` forward because we still want the old API to work during the deprecation cycle.

File: skimage/util/tests/test_compare.py
Comment (lagru) [inline skimage/util/tests/test_compare.py:32]: Might at least use `match=".*image[01]` here and elsewhere.

**Review by lagru — COMMENTED** [review]
Thanks for tackling this @mkcor. Especially the somewhat complex documentation of the deprecation. :sweat_smile:

File: skimage/util/compare.py
Comment (mkcor) [reply skimage/util/compare.py:13]: Ok. Should I write a custom wrapper then? Just to be sure, do you agree that a) I can't use `deprecate_parameter` here and b) existing code like `compare_images(image1=a, image2=b)` should still work during the deprecation cycle as well?

Comment (lagru) [reply skimage/util/compare.py:13]: a) yes and b) yes. I think a specific wrapper should be able to address all concerns in this case. I think something like
```python

def _rename_image_params(func):

    @functools.wraps(func)
    def wrapper(*args, **kwargs):
        # check / fix kwargs and warn if `image2` is used
        return func(*args, **kwargs)

    return wrapper
``` 
should work.

File: skimage/util/compare.py
Comment (lagru) [reply skimage/util/compare.py:38]: What would `change_parameter` do? Maybe for a start we can focus on this case for now after which we have a better understanding and are already part-way there if we want to create `change_parameter`; at least that's how https://github.com/scikit-image/scikit-image/pull/7256 started.

File: skimage/util/compare.py
Comment (mkcor) [reply skimage/util/compare.py]: Thank you, @lagru. I'll update the function and tests accordingly.

**Comment by mkcor** [comment]
@lagru I've drawn from your class `deprecate_parameter` but, as you will see, it's a bit ugly, hard-coded, etc. :see_no_evil: I rushed to submit this, in case it could make it into v0.23. Cc @stefanv

File: skimage/util/compare.py
Comment (lagru) [inline skimage/util/compare.py:105]: I'm sorry but unfortunately I think looking at `deprecate_parameter` has lead you on a wrong path and over-complicated things. And also breaks

```python
a = np.ones(10)
b = np.zeros(10)
ski.util.compare_images(a, b, "blend")
# warns but still uses method="diff"
```

which used to work before and should continue working during the deprecation. I think we can't use an argument with `=DEPRECATED` here.

Except for that I think you got quite close, especially with the idea to first turn all `args` into `kwargs` to more easily handle them. What do you think about the draft and decorator below? It should work for all cases unless I've forgotten a a way to abuse the signature. It also turns `method` into a keyword-only parameter.

```python
def _rename_image_params(func):

    @functools.wraps(func)
    def wrapper(*args, **kwargs):

        # Turn all into keyword parameters
        for i, (value, param) in enumerate(
            zip(args, ["image0", "image1", "method", "n_tiles"])
        ):
            if i >= 2:
                warnings.warn("method will become keyword-only")
            if param in kwargs:
                raise ValueError(
                    f"{param} given as positional and keyword argument"
                )
            else:
                kwargs[param] = value
        args = tuple()

        # Account for `image2` if given
        if "image2" in kwargs:
            warnings.warn("used deprecated image2")

            # Safely move `image2` to `image1` if that's empty
            if "image1" in kwargs:
                # Safely move `image1` to `image0`
                if "image0" in kwargs:
                    raise ValueError(
                        "three images given, `image0`, `image1` and `image2`"
                    )
                kwargs["image0"] = kwargs.pop("image1")
            kwargs["image1"] = kwargs.pop("image2")

        return func(*args, **kwargs)

    return wrapper
```

Of course it doesn't have the nice convenience stuff of modifying the docstring and warnings and error messages are drafts and need to be polished. But I think this deprecation is so special that we should just edit the docstring by hand.

File: skimage/util/compare.py
Comment (lagru) [inline skimage/util/compare.py:74]: The wrapper shouldn't recognize the `arg` for `method` by value but by position. The function will/should handle if a wrong value is passed as `method`.

File: skimage/util/compare.py
Comment (mkcor) [reply skimage/util/compare.py:105]: > I'm sorry but unfortunately I think looking at `deprecate_parameter` has lead you on a wrong path and over-complicated things.

No worries! I did this quickly and thought I would iterate afterwards.

> And also breaks
> 
> ```python
> a = np.ones(10)
> b = np.zeros(10)
> ski.util.compare_images(a, b, "blend")
> # warns but still uses method="diff"
> ```
> 
> which used to work before and should continue working during the deprecation.

Pardon me? I just tried locally and it works; it warns, indeed, and it returns
```py
array([0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5])
```
just like current main. `method="diff"` returns
 ```py
array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])
```
instead. I just noticed that `method="checkerboard"` errors because `a` and `b` are 1-dimensional, which doesn't work with
https://github.com/scikit-image/scikit-image/blob/e453a53d23a980408365d35579008a588868ffe0/skimage/util/compare.py#L48

So we should probably improve the function by checking, in the `method="checkerboard"` case, that ndim == 2 :grimacing:

File: skimage/util/tests/test_compare.py
Comment (mkcor) [reply skimage/util/tests/test_compare.py:31]: I'm not sure... I removed the case where input images are passed as kwargs following the new API (e0d5929d4e321c1ae), so we're not quite cycling over all image0-image1 combinations when expecting a warning. Or maybe my command of regular expressions is not up to the mark :wink:

File: skimage/util/compare.py
Comment (mkcor) [reply skimage/util/compare.py:105]: > So we should probably improve the function by checking, in the `method="checkerboard"` case, that ndim == 2 😬

Done db1b7740e21637a3. If you prefer, I can cherry-pick this commit and submit it as a separate PR.

File: skimage/util/compare.py
Comment (mkcor) [reply skimage/util/compare.py:74]: Yes, the function does handle it:
https://github.com/scikit-image/scikit-image/blob/e453a53d23a980408365d35579008a588868ffe0/skimage/util/compare.py#L59-L62
Basically I wanted to cover both cases, where a user might call:
```py
compare_images(a, b, 'diff') 
```
and another one might call:
```py
compare_images(a, b, c, 'diff') 
```
which should definitely error -- but this should also be taken care of by the function, right? I realize that my RuntimeError("Use `image0, image1` to pass the two input images.") isn't catching it anyway, it's only catching
```py
compare_images(image0=a, image1=b, image2=c, ...) 
```

File: skimage/util/compare.py
Comment (lagru) [reply skimage/util/compare.py:105]: Ah, you are right, it does handle this. I chose my example poorly and got confused by the `image2=DEPRECATED` as the third positional arg. 

... which brings me to the point that we should probably remove it and make it so that the function shows the expected new signature as un-ambiguously as possible.

File: skimage/util/compare.py
Comment (lagru) [reply skimage/util/compare.py:74]: That's why I think we shouldn't use the value of the 3rd and 4th argument to determine which one is `method`... just don't use `image2=DEPRECATED` and we can simply use the position reliably? Right now it seems even harder to think of all edge cases. E.g. consider
```python
import skimage as ski
a = np.ones((10, 10)) * 3
b = np.zeros((10, 10))
ski.util.compare_images(a, b, "unknown")
```
This will neither warn nor error but will simply return as if the user requested `compare_images(a, b)` (default `method=diff`). It should definitely error.

Comment (lagru) [reply skimage/util/compare.py:74]: Replying to your question: `compare_images(a, b, c, 'diff') ` should definitely error but it's not an expected use case of the old or new API so users shouldn't really encounter this case unless they try to mess around.

**Comment by lagru** [comment]
Sorry, @mkcor I don't want to come of as devaluing your approach. I'm happy with your solution as long as it gets all edge cases. :) There are a lot of edge cases to keep in mind and the longer I think about it the more I'm undecided which signature is actually least confusing to users during the deprecation.

Is it (a)
```python
def compare_images(image0, image1, image2=DEPRECATED, method='diff', *, n_tiles=(8, 8))
```
because it's easy to see that `image2` is deprecated but harder to see how `method` is affected? It's subtle to notice that not being able to use `image2` positionally also means `method` can no longer be used this way.

Or is it (b)
```python
def compare_images(image0, image1, *, method='diff', n_tiles=(8, 8))
```
because this one makes it easy to see that `method` is now a keyword-only parameter.

**Comment by mkcor** [comment]
> Sorry, @mkcor I don't want to come off as devaluing your approach.

All good, @lagru! I think my code wasn't very elegant, so I definitely welcome your inputs. :heart: 

Now, maybe this kind of code doesn't have to be the most elegant anyway, since we will remove it two releases later?

I haven't looked closely at your alternative yet, just because potential additional edge cases came up... 

> I'm happy with your solution as long as it gets all edge cases. :) There are a lot of edge cases to keep in mind

Exactly, that's the point.

> and the longer I think about it the more I'm undecided which signature is actually least confusing to users during the deprecation.
> 
> Is it (a)
> 
> ```python
> def compare_images(image0, image1, image2=DEPRECATED, method='diff', *, n_tiles=(8, 8))
> ```
> 
> because it's easy to see that `image2` is deprecated but harder to see how `method` is affected? It's subtle to notice that not being able to use `image2` positionally also means `method` can no longer be used this way.

Exactly, that's the main problem with this deprecation (more so than `image1, image2` changing to `image0, image1`).

> Or is it (b)
> 
> ```python
> def compare_images(image0, image1, *, method='diff', n_tiles=(8, 8))
> ```
> 
> because this one makes it easy to see that `method` is now a keyword-only parameter.

This was my initial move, but you said (and rightfully so) that the old API (where `method` may be used positionally) should still work during the deprecation cycle... What about (c)

```python
def compare_images(image0, image1, method='diff', *, n_tiles=(8, 8))
```
?

File: skimage/util/compare.py
Comment (mkcor) [reply skimage/util/compare.py:74]: > Replying to your question: `compare_images(a, b, c, 'diff') ` should definitely error but it's not an expected use case of the old or new API so users shouldn't really encounter this case unless they try to mess around.

Right; there's definitely no reason to 'handle' it within the function itself and perhaps not even through the decorator...

Comment (mkcor) [reply skimage/util/compare.py:74]: > That's why I think we shouldn't use the value of the 3rd and 4th argument to determine which one is `method`... just don't use `image2=DEPRECATED` and we can simply use the position reliably? Right now it seems even harder to think of all edge cases. E.g. consider
> 
> ```python
> import skimage as ski
> a = np.ones((10, 10)) * 3
> b = np.zeros((10, 10))
> ski.util.compare_images(a, b, "unknown")
> ```
> 
> This will neither warn nor error but will simply return as if the user requested `compare_images(a, b)` (default `method=diff`). It should definitely error.

Ouch, yes. Maybe we should also check the args' types then? Let me try something based on your [suggestion](https://github.com/scikit-image/scikit-image/pull/7322/files#r1533795486) and ensure that yet this other edge case is covered.

Comment (mkcor) [reply skimage/util/compare.py:74]: In the end, I was lazy and went for 9b9f757ecc619be (not trying to figure out what the third positional argument was meant to be...).

**Comment by mkcor** [comment]
Not sure we're covering all edge cases, but this implementation is definitely lighter! Thanks, @lagru. :relaxed:

File: skimage/util/compare.py
Comment (lagru) [inline skimage/util/compare.py]:
```suggestion
def compare_images(image0, image1, *, method='diff', n_tiles=(8, 8)):
```
Because the wrapper turns all given `args` into `kwargs` we can already move the asterisk (marking keyword-only arguments) to the final position.

File: skimage/util/compare.py
Comment (lagru) [inline skimage/util/compare.py]: Thanks for documenting the change in such detail! Though, what do you think about simplifying it somewhat so it's easier to grok?

```suggestion
    image0, image1 : ndarray, shape (M, N)
        Images to process, must be of the same shape.

        .. versionchanged:: 0.23
            `image1` and `image2` where renamed to `image0` and `image1`
            respecitively.
```

File: skimage/util/tests/test_compare.py
Comment (lagru) [inline skimage/util/tests/test_compare.py:20]:
```suggestion
      with pytest.raises(ValueError):
          compare_images(a, b, method="unknown")
```
I'd keep the tests focused and independent from each other; you're already testing the deprecation in another test. :)

File: skimage/util/tests/test_compare.py
Comment (lagru) [inline skimage/util/tests/test_compare.py:31]:
```suggestion
    result = compare_images(img1, img2, method='diff')
```

File: TODO.txt
Comment (lagru) [inline TODO.txt:15]:
```suggestion
```
Not needed, see other comment.

**Review by lagru — APPROVED** [review]
@mkcor thanks for addressing the discussion! I've pushed https://github.com/scikit-image/scikit-image/pull/7322/commits/0756a563c27cb8757d513025bea58192430f4f5c to correct and test the stacklevels. I'm happy with this but left a few minor concerns to make the completing the deprecation easier.

File: skimage/util/compare.py
Comment (mkcor) [reply skimage/util/compare.py]: Oh, right! Thanks, @lagru.

File: skimage/util/tests/test_compare.py
Comment (mkcor) [reply skimage/util/tests/test_compare.py:20]: Ok, you're right. I'll just edit the indentation of your edit.

**Comment by mkcor** [comment]
Thank you for the review and for 0756a563c27cb8757d513025bea58192430f4f5c, @lagru! I just included minor fixes in the docstring (770499a42a66c5058551738bc374601a52255842).

**Comment by mkcor** [comment]
@lagru thank you for taking care of 17cfe6477db9ead20d76a4c25526619874fd31aa! :pray:

File: skimage/util/compare.py
Comment (lagru) [inline skimage/util/compare.py:66]: @mkcor I corrected a minor indent issue in https://github.com/scikit-image/scikit-image/pull/7322/commits/9a10712392ba8bd2656cc47215eb4de8a1513441. Also, I'm thinking about how we can reduce the visual noise these kind of directives create in our docs. E.g. this directive I'm commenting on doesn't seem all that useful. We should probably reserve using them for breaking changes when they are absolutely necessary to avoid user confusion. And do we plan on removing them ever? Curious what you think. :)

Let's merge once the CI is green.

**Comment by lagru** [comment]
No idea about the new [error](https://github.com/scikit-image/scikit-image/actions/runs/8819647825/job/24213215782?pr=7322#step:5:208) yet:
```
../meson.build:1:0: ERROR: Compiler /usr/bin/clang cannot compile programs.
```
But it seems unrelated since this PR doesn't touch any part of the infrastructure..

**Review by jarrodmillman — APPROVED** [review]
Will merge once #7407 is sorted out and merged. Thanks!!

**Review by stefanv — APPROVED** [review]

**Pull request merged**
